apis/placement/v1beta1/clusterresourceplacement_types.go

```diff
@@ -657,53 +657,78 @@ const (
 
 	// ApplicableResourceOverrides contains a list of applicable ResourceOverride snapshots associated with the selected
 	// resources.
 	//
 	// This field is alpha-level and is for the override policy feature.
-	// +optional
+	// +kubebuilder:validation:Optional
 	ApplicableResourceOverrides []NamespacedName `json:"applicableResourceOverrides,omitempty"`
 
 	// ApplicableClusterResourceOverrides contains a list of applicable ClusterResourceOverride snapshots associated with
 	// the selected resources.
 	//
 	// This field is alpha-level and is for the override policy feature.
-	// +optional
+	// +kubebuilder:validation:Optional
 	ApplicableClusterResourceOverrides []string `json:"applicableClusterResourceOverrides,omitempty"`
 
 	// +kubebuilder:validation:MaxItems=100
 
 	// FailedPlacements is a list of all the resources failed to be placed to the given cluster or the resource is unavailable.
 	// Note that we only include 100 failed resource placements even if there are more than 100.
 	// This field is only meaningful if the `ClusterName` is not empty.
-	// +optional
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:MaxItems=100
 	FailedPlacements []FailedResourcePlacement `json:"failedPlacements,omitempty"`
 
+	// DriftedPlacements is a list of all the resources whose label/annotation and spec fields have changed on the
+	// target cluster after it is successfully applied.
+	// This field is only meaningful if the `ClusterName` is not empty.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:MaxItems=100
+	DriftedPlacements []DriftedResourcePlacement `json:"driftedPlacements,omitempty"`
+
 	// Conditions is an array of current observed conditions for ResourcePlacementStatus.
-	// +optional
+	// +kubebuilder:validation:Optional
 	Conditions []metav1.Condition `json:"conditions,omitempty"`
 }
 
 // FailedResourcePlacement contains the failure details of a failed resource placement.
 type FailedResourcePlacement struct {
 	// The resource failed to be placed.
-	// +required
+	// +kubebuilder:validation:Required
 	ResourceIdentifier `json:",inline"`
 
 	// The failed condition status.
-	// +required
+	// +kubebuilder:validation:Required
 	Condition metav1.Condition `json:"condition"`
 }
 
+// DriftedResourcePlacement contains the drifted resource details.
+type DriftedResourcePlacement struct {
+	// The resource that has drifted.
+	// +kubebuilder:validation:Required
+	ResourceIdentifier `json:",inline"`
+
+	// The ObservedGeneration of the resource on the target cluster that caused this drift.
+	// +kubebuilder:validation:Required
+	ObservedGeneration int64 `json:"observedGeneration"`
+
+	// The details of difference between the resource on the target cluster and the resource to be applied from the hub.
+	// We will also emit an event with the difference details.
+	// +kubebuilder:validation:maxLength=8192
+	// +kubebuilder:validation:Optional
+	DriftDetail string `json:"driftDetail,omitempty"`
```

Comment:
TBH I am a bit worried about the size limit as well. Should we mention directly that we might be truncating this information as necessary and the user is expected to check on their own for large diffs?